### PR TITLE
Rpc: Add getCirculatingSupply endpoint 

### LINKED ISSUE #9433: RPC: add getCirculatingSupply API
Like https://docs.solana.com/apps/jsonrpc-api#gettotalsupply but the sum of all accounts that aren’t the community/foundation/grant wallets and aren’t stake accounts with lockups.

This value could be computed the first time `getCirculatingSupply` is called, and then cached for an epoch.  It doesn't need to be 100% realtime.

The output, as of today, should be `◎ 8258353.636363637`

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Users want to know the circulating supply in the cluster; even though we know this approximate value, currently we have no way to calculate it.

#### Summary of Changes
- Add `update_supply` module to calculate circulating and non-circulating supply on epoch boundaries
- Plumb `getCirculatingSupply` into rpc, which returns circulating and non-circulating supplies, and the list of non-circulating accounts. Note: these values are static, even though total supply may fall across an epoch, since supply is updated once at the beginning of the epoch.
- On mainnet-beta, returns circulating supply: ◎19,681,955.661301762

Fixes #9433 
Closes #9541

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
@mvines , I wasn't able to put the non-circulating accounts list in the genesis crate and also add the computation to `get_entered_epoch_callback()` in genesis_programs, due to circular dependency. Alternatives to packing it all in genesis_programs would be putting it in a separate crate or something lower-level (like runtime). Thoughts?

Also, should I change the endpoint name to `getSupply` since it also returns the non-circulating balance?

**Comment by CriesofCarrots** [comment]
One other note: currently this code will only populate `Bank::supply` on the next epoch boundary. If we want it to populate sooner, we could plug in a particular slot for mainnet-beta/testnet, which should keep the data consistent across nodes.

**Comment by mvines** [comment]
When a node boots from a snapshot, it looks like it won't have a valid bank supply until the next epoch?   Call `update_supply()` after snapshot deserialization in `core/src/validator.rs?`

**Comment by mvines** [comment]
> @mvines , I wasn't able to put the non-circulating accounts list in the genesis crate and also add the computation to `get_entered_epoch_callback()` in genesis_programs, due to circular dependency. Alternatives to packing it all in genesis_programs (as I did) would be putting it in a separate crate or something lower-level (like runtime). Thoughts?

The current location in `genesis_programs` seems not bad at all


> Also, should I change the endpoint name to `getSupply` since it also returns the non-circulating balance?

Yeah I like it, which makes me want to also deprecate `getTotalSupply` and add a new total field in  `getSupply` as well

File: genesis-programs/src/non_circulating_accounts.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+pub(crate) const NON_CIRCULATING_ACCOUNTS: [&str; 25] = [
```
Comment (mvines) [inline genesis-programs/src/non_circulating_accounts.rs:1]: Some brief comments in this file that mark these accounts as mainnet-beta accounts would be nice.

File: genesis-programs/src/non_circulating_accounts.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    "9huDUZfxoJ7wGMTffUE7vh1xePqef7gyrLJu9NApncqA",
+    "GK2zqSsXLA2rwVZk347RYhh6jJpRsCA69FjLW93ZGi3B",
+    "HCV5dGFJXRrJ3jhDYA4DCeb9TEDTwGGYXtT3wHksu2Zr",
+    "25odAafVXnd63L6Hq5Cx6xGmhKqkhE2y6UrLVuqUfWZj",
+    "14FUT96s9swbmH7ZjpDvfEDywnAYy9zaNhv4xvezySGu",
```
Comment (mvines) [inline genesis-programs/src/non_circulating_accounts.rs:6]: Try `solana_sdk::declare_id!()` to produce a `Pubkey` at compile time

**Comment by CriesofCarrots** [comment]
> When a node boots from a snapshot, it looks like it won't have a valid bank supply until the next epoch? Call `update_supply()` after snapshot deserialization in `core/src/validator.rs?`

True. How to handle the fact that nodes booting from snapshots at different slots will have different circulating-supply values? Perhaps it's enough to add a `slot_when_calculated` field to the Supply struct?

**Comment by mvines** [comment]
> True. How to handle the fact that nodes booting from snapshots at different slots will have different circulating-supply values? Perhaps it's enough to add a `slot_when_calculated` field to the Supply struct?

Sure, that works.  `effective: Slot`, `updated: Slot`, _.

**Comment by CriesofCarrots** [comment]
> Yeah I like it, which makes me want to also deprecate `getTotalSupply` and add a new total field in `getSupply` as well

Right now, total supply drops constantly as fees/rent are burned. If we add total supply to this response, it seems like it probably makes the most sense to calculate/cache it once at the beginning of the epoch and hold it constant like the other values. Agree?

**Comment by mvines** [comment]
> > Yeah I like it, which makes me want to also deprecate `getTotalSupply` and add a new total field in `getSupply` as well
> 
> Right now, total supply drops constantly as fees/rent are burned. If we add total supply to this response, it seems like it probably makes the most sense to calculate/cache it once at the beginning of the epoch and hold it constant like the other values. Agree?

Oh wait.  What if we just removed all the bank-level integration, and calculated everything in the `getSupply` RPC implementation?  That would give us the current supply for all fields and keep runtime/ out of it entirely

**Comment by CriesofCarrots** [comment]
> Oh wait. What if we just removed all the bank-level integration, and calculated everything in the `getSupply` RPC implementation? That would give us the current supply for all fields and keep runtime/ out of it entirely

Do you mean abandon the epoch timing altogether, and calculate everything on every request?
We could do that. It is just a little bit expensive, since we have to iterate through all the stake accounts.

Another option is to calculate/cache the non-circulating account addresses on the epoch boundary, but get balances to calculate supply totals on rpc request.

**Comment by mvines** [comment]
> Do you mean abandon the epoch timing altogether, and calculate everything on every request?
> We could do that. It is just a little bit expensive, since we have to iterate through all the stake accounts.

Yes.  We can also cache the value within RPC too for some number of slots as well.

**Comment by CriesofCarrots** [comment]
> Yes. We can also cache the value within RPC too for some number of slots as well.

Sort of like #9541 ? 😆 

Okay, I think I'll scrap this, and PR:
- deprecate `getTotalSupply` and return as part of one `getSupply` method, wrapped in slot context
- all computation in RPC on request, no caching for now. We could add caching as a later optimization, if it seems necessary

File: genesis-programs/src/non_circulating_accounts.rs (hunk 2 above)
Comment (CriesofCarrots) [reply genesis-programs/src/non_circulating_accounts.rs:6]: Hmm, can you point me at an example that would work for this case?
I think I would need to a different macro that can handle multiple keys to produce an `ids()` method...?

Comment (mvines) [reply genesis-programs/src/non_circulating_accounts.rs:6]: You're right, it won't work out of the box.   A `pubkey!("25odAafVXnd63L6Hq5Cx6xGmhKqkhE2y6UrLVuqUfWZj")` macro that produces a Pubkey struct at compile time?

**Pull request closed without merging**